### PR TITLE
Fix timestamp on the timeseries based charts (UTC -> LOCAL TZ)

### PR DESCRIPTION
**Pull request by shmsr**
The previous behaviour on the time series based charts was that the time was represented in UTC which wasn't user-friendly.

After the fix, the timestamp is shown in the client's local timezone which is taken from the browser of the client.

Note: It also solves the problem of hive query failing for date-time columns (when in epoch_ms format) as hive doesn't support `/` operator rather it supports `DIV`

### CATEGORY

Choose one

- [x] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Screenshots are corresponding to `bintime = 1569906000` and `Timezone = Asia/Calcutta`

From epoch converter, results are:
`UTC = Tuesday, October 1, 2019 5:00:00 AM`
`IST = Tuesday, October 1, 2019 10:30:00 AM GMT+05:30`

We can hence verify the values from the chart where we have hovered on.

### Before:

Chart:
Before 1

Query:
Before 2

### After:

Chart:
After 1

Query:
After 2

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/templates/appbuilder/footer.html [hunk 1]
```diff
@@ -3,7 +3,7 @@
 {% set copyright = appbuilder.app.config.get('COPYRIGHT') %}
 <div class="navbar navbar-default navbar-fixed-bottom navbar-template navbar-default-template navbar-fixed-bottom-template">
     <div class="footer-left">
-      <label class='footer-left-label'>{{ timezone }}</label>
+      <label class='footer-left-label'>SERVER TZ: {{ timezone }}</label>
```
[the remark this answers is not in the document]
Comment (subham-sarkar-guavus) [reply superset/templates/appbuilder/footer.html]: Yes. Suggest me the name, I'll do it. Timezone mentioned there was full of conflict, whether it's the timezone where the Superset server is hosted or the timezone of the client from where superset is accessed. So added the prefix there.

File: superset/db_engine_specs.py [hunk 2]
```diff
@@ -115,16 +115,20 @@ class BaseEngineSpec(object):
...
 
     @classmethod
-    def get_time_expr(cls, expr, pdf, time_grain, grain):
+    def get_time_expr(cls, expr, pdf, time_grain, grain, timezone):
+        if timezone == None:
+            timezone = 'UTC'
         # if epoch, translate to DATE using db specific conf
         if pdf == 'epoch_s':
-            expr = cls.epoch_to_dttm().format(col=expr)
+            sub_expr = 'from_utc_timestamp(from_unixtime(%s), \'%s\')' % (expr, timezone)
```
[the remark this answers is not in the document]
Comment (subham-sarkar-guavus) [reply superset/db_engine_specs.py]: Worked for most of the SQL based DBs (Maybe all).
For example, works with, MySQL, Hive, MariaDB, Postgres, etc.

Ref: https://www.tutorialspoint.com/sql/sql-date-functions.htm

You can find all the functions I've used here, so it's not hive specifically.

Plus, consider using `DIV` and not `\` also helps as when the format of unix_timestamp is in epoch_ms we divide it by 1000 and then do the query. So using `DIV` works for all given the fact that `\` fails for Hive. So, after making the changes, we're good to go for `epoch_ms` formats as well.

File: superset/templates/appbuilder/footer.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (subham-sarkar-guavus) [reply superset/templates/appbuilder/footer.html]: Writing UTC every time is very misleading. Suppose if the server's in `America/New York` timezone, and for the calculations, it'll pick that. Although, I have pointed out changes to make it UTC always, in another PR. So, the differences of the relativeness for `epoch_s` and `epoch_ms` timestamps are taken relative from UTC which is correct for all further calculations. I've also verified the data manually from hive, and the output's satisfactory.

So, writing UTC is misleading client-side users.

File: superset/db_engine_specs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (subham-sarkar-guavus) [reply superset/db_engine_specs.py]: Also, prior to this, the `epoch_to_dttm` used to return this `from_unixtime(unix_timestamp(from_unixtime(bin_time)), 'yyyy-MM-dd HH')`, I'm just using this one, `from_unixtime(unix_timestamp(from_utc_timestamp(from_unixtime(bin_time), 'Asia/Calcutta')), 'yyyy-MM-dd HH')`. Hence, no issues.

[the remark this answers is not in the document]
Comment (subham-sarkar-guavus) [reply superset/db_engine_specs.py]: I've resolved some issues that could've broken this code. Also, as per the suggested I'm using the usual `epoch_to_dttm()` only. Hope the changes look good. They also passed all the UTs.

File: superset/templates/appbuilder/footer.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (arpit-agarwal) [reply superset/templates/appbuilder/footer.html]: Is this actually a server timezone?
If yes i would say `Application Timezone`.
However in context of other PR it become irrelevant.
Assuming both will be merged together we can discuss what should be suffix

File: superset/connectors/sqla/models.py [hunk 3]
```diff
@@ -163,8 +163,12 @@ def get_timestamp_expression(self, time_grain):
...
         expr = db.db_engine_spec.get_time_expr(
             self.expression or self.column_name,
-            pdf, time_grain, grain)
-        return literal_column(expr, type_=DateTime).label(label)
+            pdf,
+            time_grain,
+            grain,
+            timezone)
+        # Add the date-time column in select statement for data validation (epoch_s/ epoch_ms)
+        return [literal_column(expr, type_=DateTime).label(label), literal_column(self.expression or self.column_name)]
```
Comment (arpit-agarwal) [inline superset/connectors/sqla/models.py]: Is this relevant to changes for showing client timezone in timeseries

**Review by arpit-agarwal — CHANGES_REQUESTED** [review]
I need to understand overall flow and scope of this PR.
This seems to doing lot more than just timeseries time in local timezone

File: superset/db_engine_specs.py [hunk 4]
```diff
@@ -1480,7 +1487,7 @@ class PinotEngineSpec(BaseEngineSpec):
     time_grain_functions = {k: None for k in _time_grain_to_datetimeconvert.keys()}
 
     @classmethod
-    def get_time_expr(cls, expr, pdf, time_grain, grain):
+    def get_time_expr(cls, expr, pdf, time_grain, grain, timezone):
```
Comment (arpit-agarwal) [inline superset/db_engine_specs.py]: why we have added a redundant parameter

File: superset/viz.py [hunk 5]
```diff
@@ -333,6 +337,7 @@ def query_obj(self):
         }
 
         d = {
+            'timezone': client_tz,
```
Comment (arpit-agarwal) [inline superset/viz.py]: where this timzone property is used.
can you share a flow of method invocations that is stitching this e2e data flow

**Review by arpit-agarwal — COMMENTED** [review]
#DYNAMIC TZ
`select to_utc_timestamp(from_unixtime(1389802875), from_unixtime(unix_timestamp(), "z"));`

We may need this is we don't know the timezone at which `from_unixtime` will convert to

**Pull request closed without merging**